### PR TITLE
addition of houston exception and handling

### PR DESCRIPTION
**Pull request by hwindsor**
## Pull Request Overview

- Common HoustonException created for handling of errors from EDM/ACM any other parsing in the system
- Examples of where it can be raised in EDMManager and AssetGroupMetadata and how the contents can be used to generate any form of response 

Driven by comments on github.com/WildMeOrg/houston/pull/177/files
This is the  refactoring I think would make that change simpler
---

### CODE REVIEW

File: app/modules/encounters/resources.py [hunk 1]
```diff
@@ -81,11 +82,12 @@ def post(self, args):
...
-            'encounter.data', 'post', {'data': data}, ''
-        )
-        if not result_data:
-            abort(success=False, passed_message=message, message='Error', code=400)
+        try:
+            result_data = current_app.edm.request_passthrough_result(
+                'encounter.data', 'post', {'data': data}, ''
+            )
+        except HoustonException as ex:
+            abort(400, 'Error', success=False, passed_message=ex.message)
```
Comment (hwindsor) [inline app/modules/encounters/resources.py:90]: 400 only as that's what the existing code did. Is there any reason this did not use the edm status code (other than it wasn't actually available here in the existing code?

File: app/utils.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+
+class HoustonException(Exception):
+    def __init__(self, **kwargs):
+        log_message = kwargs['log_message'] if 'log_message' in kwargs else ''
+        self.message = kwargs['message'] if 'message' in kwargs else log_message
+        self.status_code = kwargs['status_code'] if 'status_code' in kwargs else 400
+
+        # Allow other params to be passed in exception
+        self._kwargs = dict(kwargs)
```
Comment (hwindsor) [inline app/utils.py:21]: feedback requested, 
I put the dict shallow copy in as I didn't know if I could safely rely on self._kwargs = kwargs

File: app/modules/encounters/resources.py (hunk 1 above)
Comment (naknomum) [reply app/modules/encounters/resources.py:90]: this ends up being kinda moot point i guess as i completely gut this `post()` part out of here via PR177!

File: app/utils.py (hunk 2 above)
Comment (naknomum) [reply app/utils.py:21]: definitely bumping up against my (shallow, much like your copy!) knowledge of python here; so not much feedback to offer.

**Review by naknomum — APPROVED** [review]
lgtm

File: app/utils.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import logging
+
+from flask_login import current_user  # NOQA
+
+log = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+class HoustonException(Exception):
+    def __init__(self, **kwargs):
+        log_message = kwargs['log_message'] if 'log_message' in kwargs else ''
```
Comment (karenc) [inline app/utils.py:16]: You can do this instead:

```python
self.message = kwargs.get('message', kwargs.get('log_message', ''))
self.status_code = kwargs.get('status_code', 400)
```

File: app/utils.py (hunk 2 above)
Comment (karenc) [reply app/utils.py:21]: I don't think we need to copy `kwargs`

File: app/utils.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        self.status_code = kwargs['status_code'] if 'status_code' in kwargs else 400
+
+        # Allow other params to be passed in exception
+        self._kwargs = dict(kwargs)
+
+        log.warning(f'Failed: {log_message} {self.status_code}')
+        # TODO This is where Audit Logging will hook into the system.
+
+    def get_string_val(self, argval):
+        return self._kwargs[argval] if argval in self._kwargs else ''
```
Comment (karenc) [inline app/utils.py:27]: Similarly you can do:

```python
return self._kwargs.get(argval, '')
```

and

```python
return self._kwargs.get(argval, 0)
```

**Review by karenc — COMMENTED** [review]
Instead of checking for whether a key exists in a dict we can do this:

```python
dict.get(key, return_this_if_key_does_not_exist)
```

File: app/utils.py (hunk 4 above)
Comment (hwindsor) [reply app/utils.py:27]: Thanks, even better I can replace these two functions with one that uses the same idea of a "not found" default. That's the only difference here.

**Pull request merged**